### PR TITLE
Convert GAS to NASM #173

### PR DESCRIPTION
**Pull request by the-grue**
#173 Conversion of GAS to NASM

Covers files in kernel/arch/i386:

- crtn.S
- crti.S
- paging_asm.S
- flush.S
- boot/boot.S
- interrupt.S

### CODE REVIEW

**Comment by Kfeavel** [comment]
Sorry for the noise, didn't see that this is a WIP until I read the description. Assign the MR to @micahswitzer and myself when you're ready for review.

File: kernel/arch/i386/interrupt.S [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+; All of the following Interrupt Service Routines are
+; exception handlers. As such, they push the error code
+; along with their exception (interrupt) number. If an
+; exception has an error code, $0 will not be pushed as
+; the CPU already does that for us.
+;
+; 0: Divide By Zero Exception
+global isr0
+isr0:
```
Comment (Kfeavel) [inline kernel/arch/i386/interrupt.S]: Is there any way to make a generic macro where all that's needed (for ISR, not exceptions) is a number, and then the following code is all macro'd away?
```asm
global isr#
isr#:
    push 0
    push #
    jmp isr_common_stub
```

File: kernel/arch/i386/interrupt.S [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+; All of the following Interrupt Requests are
+; stubs for hardware interrupts such as device interrupts.
+; These are not exceptions. Each Interrupt Request pushes
+; the IRQ value (starting at 0) along with their
+; corresponding hardware interrupt value (starting at 32).
+;
+; Interrupt Request handlers
+; System timer (PIT)
+global irq0
```
Comment (Kfeavel) [inline kernel/arch/i386/interrupt.S]: Same thing as question about the ISR, but with modifications for IRQs.

**Review by Kfeavel — CHANGES_REQUESTED** [review]
Beside my macro questions, it looks good to me. Thank you so much for contributing!

File: kernel/arch/i386/interrupt.S (hunk 1 above)
Comment (the-grue) [reply kernel/arch/i386/interrupt.S]: I'm sure there is.  I was planning to do that as a follow-on after I research nasm macros a bit.

File: kernel/arch/i386/interrupt.S (hunk 2 above)
Comment (the-grue) [reply kernel/arch/i386/interrupt.S]: Yes, both of them would be included.  Same as above.

**Review by Kfeavel — DISMISSED** [review]
All changes look good. @the-grue will open a new pull request for NASM macros later.

**Comment by the-grue** [comment]
Since we haven't merged yet, can I just add the macro refactor to this pull request?  It was easier than I had imagined.  :)

File: kernel/arch/i386/interrupt.S [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+; we have a consistent stack for all of them.
+
+; All of the following Interrupt Service Routines are
+; exception handlers. As such, they push the error code
+; along with their exception (interrupt) number. If an
+; exception has an error code, $0 will not be pushed as
+; the CPU already does that for us.
+;
+m_isr_2 0 ; 0: Divide By Zero Exception
+
```
Comment (Kfeavel) [inline kernel/arch/i386/interrupt.S]: I would remove the whitespace in between the ISRs and IRQs. It'll be nice to have more condensed source now that we have macros.

**Review by Kfeavel — CHANGES_REQUESTED** [review]
The macros look amazing. Only one nit-picky request and we'll be good to merge. 😁

File: kernel/arch/i386/interrupt.S (hunk 3 above)
Comment (the-grue) [reply kernel/arch/i386/interrupt.S]: No worries.  Just want to make sure I understand what you are asking for.  You'd like to see this:

m_isr_2 0 ; 0: Divide By Zero Exception
m_isr_2 1 ; 1: Debug Exception
m_isr_2 2 ; 2: Non Maskable Interrupt Exception
...

Comment (the-grue) [reply kernel/arch/i386/interrupt.S]: Removed whitespace to make isr / irq as above.

**Review by Kfeavel — APPROVED** [review]
Looks great!

**Pull request merged**